### PR TITLE
COMMON: Move the Lua code from SWORD25 to COMMON

### PR DESCRIPTION
**Pull request by nipunG314**
This PR moves the Lua code from the SWORD25 engine to COMMON so that Lua can be integrated into other engines.

### CODE REVIEW

**Comment by bluegr** [comment]
Thanks for this.

Is there any use for LUA scripts outside of the sword25 engine?

A lot of code has been removed from the LUA interpreter, since it is quite specific to the needs of the sword25 engine. If we need a more generic LUA interpreter, we should adapt the full original LUA code, not this engine-specific interpreter

**Comment by nipunG314** [comment]
I need the LUA scripts to write the engine for Hyperspace Delivery Boy, so they will probably be used there. If the current LUA setup is not sufficient for this, then I suppose it would make sense to adapt the full original code.

**Comment by bluegr** [comment]
Well, the original sources *have* been adapted, but some unused (in sword25) non-portable LUA functionality has been removed, or adapted. An example is the LUA I/O code. Check the history:

https://github.com/scummvm/scummvm/commits/master?after=62fc4c94bfd5b66fbfa444d686c5e86a84bea562+34&path%5B%5D=engines&path%5B%5D=sword25&path%5B%5D=util&path%5B%5D=lua

If you don’t need this functionality, then the interpreter can be used as-is, otherwise, it’ll need to be adapted for this engine. It would be best if you checked the game’s scripts for these cases.

**Comment by sev-** [comment]
Thanks.

Now, one more thing is missing. Since now the code lives in the common/, it would be great if you could turn it into an engine feature, similar to png. See sword25/configure.engine

**Comment by bluegr** [comment]
Any news on this?

**Comment by bluegr** [comment]
Any news on this? Now there are conflicts because of changes in PR #1769

**Comment by nipunG314** [comment]
@bluegr Sorry, got seriously sidetracked on this. I have rebased the breaking changes. Will work on the Engine Feature idea next

File: configure [hunk 1]
```diff
@@ -3789,6 +3794,26 @@ case $_host in
 		;;
 esac
 
+#
+# Enable Lua only for engines that support it
```
Comment (bluegr) [inline configure]: “Enable LUA only for engines that require it”

Comment (Kawa-oneechan) [reply configure]: "Lua" is actually correct. It's not an acronym.

Comment (bluegr) [reply configure]: You are quite correct, my mistake. It's "Lua"

File: configure [hunk 2]
```diff
@@ -1030,6 +1032,7 @@ Optional Features:
   --disable-nuked-opl      don't build Nuked OPL driver
   --disable-16bit          don't enable 16bit color support
   --disable-highres        don't enable support for high resolution engines >320x240
+  --disable-lua            don't enable support Lua
```
[the remark this answers is not in the document]
Comment (bluegr) [reply configure]: Please disregard this one. Your comment is correct, the proper naming is "Lua"

**Comment by nipunG314** [comment]
create_msvc wasn't detecting the Lua engine feature. I have re-worked the feature as a default option, and added a Feature Flag for it.

File: configure [hunk 3]
```diff
@@ -5397,8 +5405,8 @@ if test "$_build_scalers" = yes ; then
 	fi
 fi
 
-if test "$_mt32emu" = yes ; then
-	echo_n ", MT-32 emulator"
```
Comment (bluegr) [inline configure]: I believe that these have been removed by mistake

File: common/module.mk [hunk 4]
```diff
@@ -19,6 +19,39 @@ MODULE_OBJS := \
...
+	lua/lstrlib.o \
+	lua/ltable.o \
+	lua/ltablib.o \
+	lua/ltm.o \
+	lua/lua_persist.o \
+	lua/lua_persistence_util.o \
+	lua/lua_unpersist.o \
+	lua/lvm.o \
+	lua/lzio.o \
+	lua/scummvm_file.o \
```
Comment (bluegr) [inline common/module.mk]: These should not be compiled in all the time in common code, which is why the feature flag has been requested. They should be enclosed in a ENABLE_LUA ifdef. i.e:

    ifdef ENABLE_LUA
    MODULE_OBJS += \
	xxx.o
    endif

Check common/module.mk for an example

File: configure (hunk 3 above)
Comment (nipunG314) [reply configure]: Re-added

File: common/module.mk (hunk 4 above)
Comment (nipunG314) [reply common/module.mk]: I have wrapped it now

**Comment by nipunG314** [comment]
Any other changes required?

**Comment by sev-** [comment]
Great, merging!

**Pull request merged**